### PR TITLE
[SPARK-31892][SQL][FOLLOWUP][test-java11] Improve test coverage for datetime pattern with formatter functions 

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
This PR is a followup for https://github.com/apache/spark/pull/28706, we want fully check for the formatter, which can warn us ahead if there is any behavior change when we update JDK.

### Why are the changes needed?

test coverage and hotfix for the silly mistake in https://github.com/apache/spark/pull/28706 at https://github.com/apache/spark/pull/28718/files#diff-7f589e01d3e5e5ea284c1622527d4984R49

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

add tests

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/datetime.sql.out [hunk 1]
```diff
@@ -1013,3 +1013,110 @@ select date_format('2020-01-01', 'YYYY-MM-dd uuuu')
...
+7 07 7 07 7 07 7 07 3 03 33 33 123
+9 09 9 09 9 09 9 09 33 33 33 33 123
+
+
+-- !query
+select date_format(t, 'VV z zz zzz zzzz O OOOO X XX XXX XXXX XXXXX x xx xxx xxxx xxxx xxxxx Z ZZ ZZZ ZZZZ ZZZZZ') from ttt
+-- !query schema
+struct<date_format(t, VV z zz zzz zzzz O OOOO X XX XXX XXXX XXXXX x xx xxx xxxx xxxx xxxxx Z ZZ ZZZ ZZZZ ZZZZZ):string>
+-- !query output
+America/Los_Angeles PST PST PST Pacific Standard Time GMT-7:52:58 GMT-07:52:58 -0752 -0752 -07:52 -075258 -07:52:58 -0752 -0752 -07:52 -075258 -075258 -07:52:58 -0752 -0752 -0752 GMT-07:52:58 -07:52:58
```
Comment (yaooqinn) [inline sql/core/src/test/resources/sql-tests/results/datetime.sql.out:1080]: Do you know what causes this inaccuracy for dates before 1582-10-15？@cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala [hunk 2]
```diff
@@ -23,43 +23,49 @@ import org.apache.spark.sql.catalyst.util.DateTimeFormatterHelper._
 class DateTimeFormatterHelperSuite extends SparkFunSuite {
 
   test("check incompatible pattern") {
-    assert(convertIncompatiblePattern("MM-DD-u") === "MM-DD-e")
-    assert(convertIncompatiblePattern("yyyy-MM-dd'T'HH:mm:ss.SSSz")
-      === "uuuu-MM-dd'T'HH:mm:ss.SSSz")
-    assert(convertIncompatiblePattern("yyyy-MM'y contains in quoted text'HH:mm:ss")
-      === "uuuu-MM'y contains in quoted text'HH:mm:ss")
-    assert(convertIncompatiblePattern("yyyy-MM-dd-u'T'HH:mm:ss.SSSz")
+    assert(convertIncompatiblePattern("MM-DD-u", isParsing = false) === "MM-DD-e")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala:26]: In this test suite, how about we add
```
def convertIncompatiblePattern(pattern: String): String = {
  convertIncompatiblePattern(pattern, isParsing = false)
}
```
To reduce the code diff?

File: sql/core/src/test/resources/sql-tests/results/datetime-legacy.sql.out [hunk 3]
```diff
@@ -996,3 +996,72 @@ select date_format('2020-01-01', 'YYYY-MM-dd uuuu')
...
+1969-12-31 15:0:0 1969-12-31 15:00:00.000 69 1969 01969 Dec December 12 12 5 3 03 15 15 3 03 PM
+1970-12-31 4:59:59 1970-12-31 04:59:59.999 70 1970 01970 Dec December 12 12 5 4 04 4 04 4 04 AM
+1996-3-31 7:3:33 1996-03-31 07:03:33.123 96 1996 01996 Mar March 3 03 5 7 07 7 07 7 07 AM
+2018-11-17 5:33:33 2018-11-17 05:33:33.123 18 2018 02018 Nov November 11 11 3 5 05 5 05 5 05 AM
+2019-12-31 9:33:33 2019-12-31 09:33:33.123 19 2019 02019 Dec December 12 12 5 9 09 9 09 9 09 AM
+2100-1-1 1:33:33 2100-01-01 01:33:33.123 00 2100 02100 Jan January 1 01 1 1 01 1 01 1 01 AM
+
+
+-- !query
+select date_format(t, 'VV z zz zzz zzzz O OOOO X XX XXX XXXX XXXXX x xx xxx xxxx xxxx xxxxx Z ZZ ZZZ ZZZZ ZZZZZ') from ttt
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/datetime-legacy.sql.out:1054]: can we test `VV` separately? otherwise all other patterns listed here can't be tested with the legacy formatter.

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 4]
```diff
@@ -168,3 +168,18 @@ select date_format(date '2018-11-17', 'yyyyyyyyyyy-MM-dd');
...
+-- valid formatter pattern check
+create temporary view ttt as select t from VALUES
+ (timestamp '1582-06-01 11:33:33.123UTC+080000'),
+ (timestamp '1970-01-01 00:00:00.000Europe/Paris'),
+ (timestamp '1970-12-31 23:59:59.999Asia/Srednekolymsk'),
+ (timestamp '1996-04-01 00:33:33.123Australia/Darwin'),
+ (timestamp '2018-11-17 13:33:33.123Z'),
+ (timestamp '2020-01-01 01:33:33.123Asia/Shanghai'),
+ (timestamp '2100-01-01 01:33:33.123America/Los_Angeles') tt(t);
+select date_format(t, 'Y-w-u YYYY-ww-uu YYY-W-uuu YY YYYYY uuuu E EE EEE EEEE') from ttt;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql:181]: it's better to have a column to show the value of `t` as well.

File: sql/core/src/test/resources/sql-tests/results/datetime.sql.out (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/datetime.sql.out:1080]: I'm not sure, but it seems like we should ban all the timezone letters during formatting. Spark's timestamp value doesn't have an associated timezone, it makes no sense to extract timezone from it.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/datetime.sql.out:1080]: cc @MaxGekk

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/results/datetime.sql.out:1080]: It seems to only display the value of `spark.sql.session.timeZone`

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/datetime.sql.out:1080]: actually, this is expected. The Spark test uses timezone `America/Los_Angeles`, which is indeed GMT-7:52:58 long time before, see https://www.politico.com/story/2017/11/18/railroads-initiative-synchronizes-us-clocks-nov-18-1883-244935

Spark's timestamp value doesn't have an associated timezone, so extracting timezone means showing the session local timezone.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala [hunk 5]
```diff
@@ -42,7 +46,7 @@ class DateTimeFormatterHelperSuite extends SparkFunSuite {
     }
     unsupportedLettersForParsing.foreach { l =>
       val e = intercept[IllegalArgumentException] {
-        convertIncompatiblePattern(s"$l", isParsing = true)
+        DateTimeFormatterHelper.convertIncompatiblePattern(s"$l", isParsing = true)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala:49]: is this change needed?

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
+--- TESTS FOR DATETIME FORMATTING FUNCTIONS ---
+
+-- valid formatter pattern check
+create temporary view ttt as select t from VALUES
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting.sql:4]: nit: 
```
create temporary view v as select col from ...
t(col)
```

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/datetime-formatting.sql:4]: and please lower case `VALUES`, or upper case all SQL keywords.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala (hunk 5 above)
Comment (yaooqinn) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelperSuite.scala:49]: Yes, the compiler is unhappy w/o this change.

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+create temporary view v as select col from values
+ (timestamp '1582-06-01 11:33:33.123UTC+080000'),
+ (timestamp '1970-01-01 00:00:00.000Europe/Paris'),
+ (timestamp '1970-12-31 23:59:59.999Asia/Srednekolymsk'),
+ (timestamp '1996-04-01 00:33:33.123Australia/Darwin'),
+ (timestamp '2018-11-17 13:33:33.123Z'),
+ (timestamp '2020-01-01 01:33:33.123Asia/Shanghai'),
+ (timestamp '2100-01-01 01:33:33.123America/Los_Angeles') t(col);
+
+select col, date_format(col, 'GGGGG') from v;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql:14]: for invalid testing, we can just use literals?

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql [hunk 8]
```diff
@@ -0,0 +1,36 @@
+--- TESTS FOR DATETIME FORMATTING FUNCTIONS WITH INVALID PATTERNS ---
+
+-- separating this from datetime-formatting.sql ,because the text form
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql:3]: `.sql ,because` -> `.sql, because`

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+select date_format('2018-11-17 13:33:33.333', 'QQQQQ');
+select date_format('2018-11-17 13:33:33.333', 'MMMMM');
+select date_format('2018-11-17 13:33:33.333', 'LLLLL');
+select date_format('2018-11-17 13:33:33.333', 'www');
+select date_format('2018-11-17 13:33:33.333', 'WW');
+select date_format('2018-11-17 13:33:33.333', 'uuuuu');
+select date_format('2018-11-17 13:33:33.333', 'EEEEE');
+select date_format('2018-11-17 13:33:33.333', 'FF');
+select date_format('2018-11-17 13:33:33.333', 'ddd');
+-- DD is invalid in 8, but valid in 11 for the JDKs that PR builder uses
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting-invalid.sql:20]: DD is invalid if the day-of-year exceeds 100, but it becomes valid in Java 11.

File: sql/core/src/test/resources/sql-tests/inputs/datetime-formatting.sql [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+select col, date_format(col, 'K KK') from v;
+
+select col, date_format(col, 'm mm') from v;
+
+select col, date_format(col, 's ss') from v;
+
+select col, date_format(col, 'S SS SSS SSSS SSSSS SSSSSS SSSSSSS SSSSSSSS SSSSSSSSS') from v;
+
+-- add upper function here to avoid
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime-formatting.sql:55]: to avoid what?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@dongjoon-hyun do you know how to trigger java 11 test in the PR builder?

**Comment by MaxGekk** [comment]
@cloud-fan See the example https://github.com/apache/spark/pull/28463 - add `[test-java11]`

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Sorry for being late. Yes, @MaxGekk 's comment is correct.
